misc/python/materialize/version_list.py

```diff
@@ -0,0 +1,109 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+
+import os
+from pathlib import Path
+from typing import List
+
+import frontmatter
+
+from materialize.git import get_version_tags
+from materialize.util import MzVersion
+
+ROOT = Path(os.environ["MZ_ROOT"])
+
+
+class VersionList:
+    def __init__(self) -> None:
+        self.versions: List[MzVersion]
+        assert False
+
+    def all_versions(self) -> List[MzVersion]:
+        return self.versions
+
+    def minor_versions(self) -> List[MzVersion]:
+        """Return the latest patch version for every minor version."""
+        minor_versions = {}
+        for version in self.versions:
+            minor_versions[f"{version.major}.{version.minor}"] = version
+
+        assert len(minor_versions) > 0
+        return sorted(minor_versions.values())
+
+    def patch_versions(self, minor_version: MzVersion) -> List[MzVersion]:
+        """Return all patch versions within the given minor version."""
+        patch_versions = []
+        for version in self.versions:
+            if (
+                version.major == minor_version.major
+                and version.minor == minor_version.minor
+            ):
+                patch_versions.append(version)
+
+        assert len(patch_versions) > 0
+        return sorted(patch_versions)
+
+
+class VersionsFromGit(VersionList):
+    """Materialize versions as tagged in Git.
+
+    >>> len(VersionsFromGit().all_versions()) > 0
+    True
+
+    >>> len(VersionsFromGit().minor_versions()) > 0
+    True
+
+    >>> len(VersionsFromGit().patch_versions(minor_version=MzVersion.parse("0.52.0")))
+    5
+
+    >>> min(VersionsFromGit().all_versions())
+    MzVersion(major=0, minor=1, patch=0, prerelease='rc', build=None)
```

Comment:
These are just part of the comment and don't get exercised, right?

Reply:
Those are doctests, that is, unit tests in actual python code that are exercised by

```
bin/pyactivate -m pytest -qq --doctest-modules misc/python
```

which is a part of 

```
bin/pycheck
```

and thus the CI pipeline .  I did not want to write doctests initially, I wanted pytest tests,  but doctests are currently the only type of python unit tests that would get run without adding more steps to the CI pipeline.